core/Pos/Binary/Class/Primitive.hs

```diff
@@ -240,3 +246,25 @@ decodeUnknownCborDataItem :: D.Decoder s ByteString
 decodeUnknownCborDataItem = do
     decodeCborDataItemTag
     D.decodeBytes
+
+-- | Encodes a type `a` , protecting it from tampering/network-transport-alteration by
+-- protecting it with a CRC.
+encodeCrcProtected :: Bi a => a -> E.Encoding
+encodeCrcProtected x = E.encodeListLen 2 <> encodeUnknownCborDataItem body <> encode (crc32 body)
+  where
+    body = serialize' x
+
+-- | Decodes a CBOR blob into a type `a`, checking the serialised CRC corresponds to the computed one.
+decodeCrcProtected :: forall s a. Bi a => D.Decoder s a
+decodeCrcProtected = do
+    enforceSize ("decodeCrcProtected: " <> show (typeOf (Proxy @a))) 2
+    body <- decodeUnknownCborDataItem
+    expectedCrc  <- decode
+    let actualCrc = crc32 body
```

Comment:
Can we get a type sig on this since currently the only thing that determines the type we encode is the return type of the `crc32` function, and if that ever changed from Int to Word or Word64 or whatever then we'd break things. Just give an explicit type:

    let actualCrc :: ???
        actualCrc = crc32 body